### ISSUE TITLE
Add specific bundle configuration to quote

### ISSUE THREAD

**Issue by jonathanwiesel** [sseixas/CPQ-JS#2]
Suppose I have a bundle "A" with 2 possible product options "B" and "C".

I want to use the `quote.addProducts` API call to add the "A" bundle with the "B" option but I'm not able to add the "B" option to the quote, just the "A" bundle. 

In the context, I'm passing a `product` model for the bundle "A" and in the `configuration` key of the model I'm adding the configuration model for the option "B" in the `optionConfigurations` array. Yet it seems the API does not take that in consideration since only adds the bundle without the option to the quote.

**Comment by sseixas** [comment]
My advice to you is to update option B to be "selected" by default in your org. Then read the product via the the `product.read` API and copy the value of the response's `configuration` property. This is the portion of the product model that when used as an input to `quote.addProducts` actually adds the product and its configuration (i.e. options). If you cannot leave option B as "selected" in the org, then you can programmatically add this `configuration` value to the product A after reading it and and before using it to add products. Make sense?

**Comment by jonathanwiesel** [comment]
@sseixas thank you, the issue at the end was that I wasn't setting the `configured = true` in the bundle's configuration.

**Comment by paustint** [comment]
@jonathanwiesel @sseixas - Hey guys, sorry for using this as a support channel - but the documentation does not really cover this use-case.

I am trying to amend a contract and "clone" a bundle so I can do a price change and am having a difficult time getting the options selected/added.

I am copying options from the `options` array to the `optionConfigurations[]`, but they are ignored and not added.

Any thoughts on what I might be doing wrong?

# Initial data obtained from load product
```javascript
{
    "virtualFeatureTitle": null,
    "upgradedAssetId": null,
    "upgradeableAssets": null,
    "relatedContent": null,
    "record": { ... },
    "productKey": null,
    "productFieldSets": null,
    "options": [
        {
            "upgradeSourcesBySourceProductId": null,
            "sourceProductIds": null,
            "record": {
                "attributes": {
                    "type": "SBQQ__ProductOption__c",
                    "url": "/services/data/v41.0/sobjects/SBQQ__ProductOption__c/a3B0q0000008ddrEAA"
                },
                "SBQQ__ConfiguredSKU__c": "01t0q000000ezfNAAQ",
                "Id": "a3B0q0000008ddrEAA",
                "OwnerId": "0050q000000WFonAAG",
                "IsDeleted": false,
                "Name": "PO-000053",
                "CurrencyIsoCode": "USD",
                "CreatedDate": "2018-01-05T15:54:22.000+0000",
                "CreatedById": "0050q000000WFonAAG",
                "LastModifiedDate": "2018-01-05T16:17:08.000+0000",
                "LastModifiedById": "0050q000000WFonAAG",
                "SystemModstamp": "2018-01-06T11:31:42.000+0000",
                "SBQQ__AppliedImmediately__c": false,
                "SBQQ__Bundled__c": false,
                "SBQQ__DiscountedByPackage__c": false,
                "SBQQ__Feature__c": "a3A0q0000008WNQEA2",
                "SBQQ__Number__c": 10,
                "SBQQ__OptionalSKU__c": "01t0q000000fxphAAA",
                "SBQQ__PriceEditable__c": "No",
                "SBQQ__ProductCode__c": "Verification - Overage",
                "SBQQ__ProductDescription__c": "atg",
                "SBQQ__ProductName__c": "Verification - Overage",
                "SBQQ__ProductSubscriptionPricing__c": "Fixed Price",
                "SBQQ__QuantityEditable__c": false,
                "SBQQ__Quantity__c": 1.00,
                "SBQQ__Required__c": false,
                "SBQQ__Selected__c": false,
                "SBQQ__System__c": false,
                "SBQQ__Type__c": "Related Product",
                "SBQQ__UpliftedByPackage__c": false,
                "Quantity_Impressions__c": 1,
                "SBQQ__UnitPrice__c": 0.00
            },
            "reconfigureDimensionWarning": "Quantities may only be adjusted from within the configurator during initial configuration.",
            "quantityEditable": false,
            "productQuantityScale": null,
            "priorOptionExists": null,
            "priceEditable": false,
            "parentProductKey": null,
            "original": null,
            "localizationByFieldName": null,
            "isUpgrade": false,
            "hasDimension": false,
            "externalConfigurationData": null,
            "exclusionGroups": {},
            "dynamicOptionKey": null,
            "dependentIds": [],
            "controllingGroups": {},
            "configurationRequired": false,
            "configurable": false
        },
        { ... ETC ...}
    ],
    "optionFieldSets": null,
    "localizationByFieldName": null,
    "localizationByFieldByProductId": null,
    "localizationByFieldByDimensionId": null,
    "inheritedConfigurationAttributes": null,
    "features": [ ... ],
    "featureCategoryLabels": {
        "Reporting": "Reporting",
        "Implementation": "Implementation",
        "Software": "Software",
        "Hardware": "Hardware"
    },
    "featureCategories": [
        "Implementation",
        "Reporting"
    ],
    "discountSchedules": null,
    "dimensions": null,
    "currencySymbol": "USD ",
    "currencyCode": "USD",
    "costs": null,
    "constraints": [],
    "configurationRules": null,
    "configurationAttributes": [],
    "configuration": {
        "validationMessages": [],
        "upgradedAssetId": null,
        "sourceLineKey": null,
        "searchFilters": null,
        "priorBundledQuantity": null,
        "priceEditable": false,
        "optionId": null,
        "optionData": {
            "attributes": {
                "type": "SBQQ__ProductOption__c"
            }
        },
        "optionConfigurations": [],
        "listPrice": null,
        "lineItemKey": null,
        "lineItemId": null,
        "isUpgrade": false,
        "isSegmented": null,
        "isDynamicOption": false,
        "inheritedConfigurationData": null,
        "hiddenOptionIds": null,
        "guidedMessages": null,
        "externalConfigurationData": null,
        "dynamicOptionKey": null,
        "disabledOptionIds": null,
        "configuredProductKey": null,
        "configuredProductId": "01t0q000000ezfNAAQ",
        "configured": false,
        "configurationLabels": null,
        "configurationEntered": false,
        "configurationData": {
            "attributes": {
                "type": "SBQQ__ProductOption__c"
            }
        },
        "changedByProductActions": false,
        "allowAssetRefund": null
    },
    "blockPrices": null
}
```

# My attempt to select the options (ProductAddContext)
```javascript
{
    "quote": {
        "record": { ... },
        "nextKey": 7,
        "netTotal": null,
        "netNonSegmentTotal": null,
        "lineItems": [ ... ],
        "lineItemGroups": [],
        "customerTotal": null,
        "channelDiscountsOffList": false,
        "calculationRequired": false,
        "applyPartnerDiscountFirst": true,
        "applyAdditionalDiscountLast": false
    },
    "products": [
        {
            "upgradedAssetId": null,
            "record": { ... },
            "options": [ ... SAME AS ABOVE ... ],
            "inheritedConfigurationAttributes": null,
            "features": [ ... ],
            "featureCategories": [
                "Implementation",
                "Reporting"
            ],
            "currencySymbol": "USD ",
            "currencyCode": "USD",
            "constraints": [],
            "configurationAttributes": [],
            "configuration": {
                "validationMessages": [],
                "priceEditable": false,
                "optionId": null,
                "optionData": {
                    "attributes": {
                        "type": "SBQQ__ProductOption__c"
                    }
                },
                "optionConfigurations": [
                    {
                        "validationMessages": [],
                        "priceEditable": false,
                        "optionId": "a3B0q0000008ddSEAQ",
                        "optionData": {
                            "attributes": {
                                "type": "SBQQ__ProductOption__c",
                                "url": "/services/data/v41.0/sobjects/SBQQ__ProductOption__c/a3B0q0000008ddSEAQ"
                            },
                            "LastModifiedDate": "2018-01-09T20:28:46.000+0000",
                            "SBQQ__Quantity__c": 1.0,
                            "SBQQ__UnitPrice__c": 0.0,
                            "SBQQ__AppliedImmediately__c": false,
                            "SBQQ__ProductDescription__c": "atg",
                            "SBQQ__QuantityEditable__c": true,
                            "SBQQ__DiscountedByPackage__c": false,
                            "Name": "PO-000049",
                            "SBQQ__Type__c": "Related Product",
                            "SBQQ__ConfiguredSKU__c": "01t0q000000ezfNAAQ",
                            "SBQQ__ProductName__c": "Tag",
                            "OwnerId": "0050q000000WFonAAG",
                            "CreatedById": "0050q000000WFonAAG",
                            "SBQQ__Required__c": false,
                            "SBQQ__Selected__c": true,
                            "SBQQ__PriceEditable__c": "No",
                            "SBQQ__ProductSubscriptionPricing__c": "Fixed Price",
                            "IsDeleted": false,
                            "SBQQ__Feature__c": "a3A0q0000008WNQEA2",
                            "SBQQ__UpliftedByPackage__c": false,
                            "SBQQ__Bundled__c": false,
                            "CurrencyIsoCode": "USD",
                            "SBQQ__Number__c": 20.0,
                            "Pricing_Method__c": "A La Carte",
                            "SystemModstamp": "2018-01-10T11:32:58.000+0000",
                            "Quantity_Impressions__c": 1.0,
                            "SBQQ__ProductConfigurationType__c": "Allowed",
                            "CreatedDate": "2018-01-04T00:46:57.000+0000",
                            "Id": "a3B0q0000008ddSEAQ",
                            "SBQQ__OptionalSKU__c": "01t0q000000fuYOAAY",
                            "LastModifiedById": "0050q000000WFonAAG",
                            "SBQQ__ProductCode__c": "Tag-Transactional",
                            "SBQQ__System__c": false
                        },
                        "optionConfigurations": null,
                        "listPrice": null,
                        "isUpgrade": false,
                        "isDynamicOption": false,
                        "inheritedConfigurationData": null,
                        "hiddenOptionIds": [],
                        "dynamicOptionKey": null,
                        "disabledOptionIds": [],
                        "configuredProductId": "01t0q000000ezfNAAQ",
                        "configured": true,
                        "configurationData": null,
                        "changedByProductActions": false
                    }
                ],
                "listPrice": null,
                "isUpgrade": false,
                "isDynamicOption": false,
                "inheritedConfigurationData": null,
                "hiddenOptionIds": null,
                "dynamicOptionKey": null,
                "disabledOptionIds": null,
                "configuredProductId": "01t0q000000ezfNAAQ",
                "configured": true,
                "configurationData": {
                    "attributes": {
                        "type": "SBQQ__ProductOption__c"
                    }
                },
                "changedByProductActions": false
            }
        }
    ],
    "ignoreCalculate": true,
    "groupKey": null
}
```

**Comment by jonathanwiesel** [comment]
@paustint believe it or not, even though @sseixas will probably kill us for spamming the repo, this 2 issues and the people involved have been the most useful piece of information I've encountered so far and I really thank you all for that.

Now about the issue, I think the problem is in the `ProductAddContext`, specifically in `products[0].configuration. optionConfigurations[0].configuredProductId` there you must specify the `SBQQ__OptionalSKU__c` for the option, not the `SBQQ__ConfiguredSKU__c`.

To reach this conclusion, I tested updating a product option and setting the `SBQQ__Selected__c` field to `true`, this way when you call the `ProductLoader` API, the `configuration.optionConfigurations` array will already have that option added. 

I hope it helps.

**Comment by paustint** [comment]
@jonathanwiesel - thanks for the info!

I am getting a null pointer exception when I set `configured = true` on the outermost `configuration` object (not the one in `optionConfigurations`).

I am leaving all the others as default values from the product load - any thoughts?

**Comment by paustint** [comment]
The first two I set to "selected by default", these get added. The third one I am adding by hand and it is ignored.

The only difference I can spot is `optionConfigurations` being null vs `[]`

# Example
```
"configuration": {
    "validationMessages": [],
    "priceEditable": false,
    "optionId": null,
    "optionData": {
        "attributes": {
            "type": "SBQQ__ProductOption__c"
        }
    },
    "optionConfigurations": [
        {
            "validationMessages": [],
            "priceEditable": false,
            "optionId": "a3B0q0000008dMPEAY",
            "optionData": {
                "attributes": {
                    "type": "SBQQ__ProductOption__c",
                    "url": "/services/data/v41.0/sobjects/SBQQ__ProductOption__c/a3B0q0000008dMPEAY"
                },
                "SBQQ__Quantity__c": 1.0,
                "SBQQ__AppliedImmediately__c": false,
                "SBQQ__ProductDescription__c": "atg",
                "SBQQ__QuantityEditable__c": true,
                "SBQQ__DiscountedByPackage__c": false,
                "SBQQ__Type__c": "Related Product",
                "SBQQ__ConfiguredSKU__c": "01t0q000000ezfNAAQ",
                "SBQQ__ProductName__c": "CM - Display",
                "OwnerId": "0050q000000WFosAAG",
                "SBQQ__Required__c": false,
                "SBQQ__Selected__c": true,
                "SBQQ__PriceEditable__c": "No",
                "SBQQ__ProductSubscriptionPricing__c": "Fixed Price",
                "IsDeleted": false,
                "SBQQ__Feature__c": "a3A0q0000008WNaEAM",
                "SBQQ__UpliftedByPackage__c": false,
                "SBQQ__Bundled__c": false,
                "CurrencyIsoCode": "USD",
                "SBQQ__Number__c": 60.0,
                "SystemModstamp": "2018-01-10T18:48:16.000+0000",
                "Quantity_Impressions__c": 1.0,
                "Id": "a3B0q0000008dMPEAY",
                "SBQQ__OptionalSKU__c": "01t0q000000fJTNAA2",
                "SBQQ__ProductCode__c": "CM - Display",
                "SBQQ__System__c": false
            },
            "optionConfigurations": [],
            "listPrice": null,
            "isUpgrade": false,
            "isDynamicOption": false,
            "inheritedConfigurationData": null,
            "hiddenOptionIds": null,
            "dynamicOptionKey": null,
            "disabledOptionIds": null,
            "configuredProductId": "01t0q000000fJTNAA2",
            "configured": false,
            "configurationData": {
                "attributes": {
                    "type": "SBQQ__ProductOption__c"
                }
            },
            "changedByProductActions": false
        },
        {
            "validationMessages": [],
            "priceEditable": false,
            "optionId": "a3B0q0000008dMUEAY",
            "optionData": {
                "attributes": {
                    "type": "SBQQ__ProductOption__c",
                    "url": "/services/data/v41.0/sobjects/SBQQ__ProductOption__c/a3B0q0000008dMUEAY"
                },
                "SBQQ__Quantity__c": 1.0,
                "SBQQ__AppliedImmediately__c": false,
                "SBQQ__ProductDescription__c": "atg",
                "SBQQ__QuantityEditable__c": true,
                "SBQQ__DiscountedByPackage__c": false,
                "SBQQ__Type__c": "Related Product",
                "SBQQ__ConfiguredSKU__c": "01t0q000000ezfNAAQ",
                "SBQQ__ProductName__c": "CM - Video",
                "OwnerId": "0050q000000WFosAAG",
                "SBQQ__Required__c": false,
                "SBQQ__Selected__c": true,
                "SBQQ__PriceEditable__c": "No",
                "SBQQ__ProductSubscriptionPricing__c": "Fixed Price",
                "IsDeleted": false,
                "SBQQ__Feature__c": "a3A0q0000008WNaEAM",
                "SBQQ__UpliftedByPackage__c": false,
                "SBQQ__Bundled__c": false,
                "CurrencyIsoCode": "USD",
                "SBQQ__Number__c": 70.0,
                "SystemModstamp": "2018-01-10T18:48:22.000+0000",
                "Quantity_Impressions__c": 1.0,
                "Id": "a3B0q0000008dMUEAY",
                "SBQQ__OptionalSKU__c": "01t0q000000fJTSAA2",
                "SBQQ__ProductCode__c": "CM - Video",
                "SBQQ__System__c": false
            },
            "optionConfigurations": [],
            "listPrice": null,
            "isUpgrade": false,
            "isDynamicOption": false,
            "inheritedConfigurationData": null,
            "hiddenOptionIds": null,
            "dynamicOptionKey": null,
            "disabledOptionIds": null,
            "configuredProductId": "01t0q000000fJTSAA2",
            "configured": false,
            "configurationData": {
                "attributes": {
                    "type": "SBQQ__ProductOption__c"
                }
            },
            "changedByProductActions": false
        },
        {
            "validationMessages": [],
            "priceEditable": false,
            "optionId": "a3B0q0000008ddSEAQ",
            "optionData": {
                "attributes": {
                    "type": "SBQQ__ProductOption__c",
                    "url": "/services/data/v41.0/sobjects/SBQQ__ProductOption__c/a3B0q0000008ddSEAQ"
                },
                "LastModifiedDate": "2018-01-09T20:28:46.000+0000",
                "SBQQ__Quantity__c": 1.0,
                "SBQQ__UnitPrice__c": 0.0,
                "SBQQ__AppliedImmediately__c": false,
                "SBQQ__ProductDescription__c": "atg",
                "SBQQ__QuantityEditable__c": true,
                "SBQQ__DiscountedByPackage__c": false,
                "Name": "PO-000049",
                "SBQQ__Type__c": "Related Product",
                "SBQQ__ConfiguredSKU__c": "01t0q000000ezfNAAQ",
                "SBQQ__ProductName__c": "Tag",
                "OwnerId": "0050q000000WFonAAG",
                "CreatedById": "0050q000000WFonAAG",
                "SBQQ__Required__c": false,
                "SBQQ__Selected__c": true,
                "SBQQ__PriceEditable__c": "No",
                "SBQQ__ProductSubscriptionPricing__c": "Fixed Price",
                "IsDeleted": false,
                "SBQQ__Feature__c": "a3A0q0000008WNQEA2",
                "SBQQ__UpliftedByPackage__c": false,
                "SBQQ__Bundled__c": false,
                "CurrencyIsoCode": "USD",
                "SBQQ__Number__c": 20.0,
                "Pricing_Method__c": "A La Carte",
                "SystemModstamp": "2018-01-10T11:32:58.000+0000",
                "Quantity_Impressions__c": 1.0,
                "SBQQ__ProductConfigurationType__c": "Allowed",
                "CreatedDate": "2018-01-04T00:46:57.000+0000",
                "Id": "a3B0q0000008ddSEAQ",
                "SBQQ__OptionalSKU__c": "01t0q000000fuYOAAY",
                "LastModifiedById": "0050q000000WFonAAG",
                "SBQQ__ProductCode__c": "Tag-Transactional",
                "SBQQ__System__c": false
            },
            "optionConfigurations": null,
            "listPrice": null,
            "isUpgrade": false,
            "isDynamicOption": false,
            "inheritedConfigurationData": null,
            "hiddenOptionIds": [],
            "dynamicOptionKey": null,
            "disabledOptionIds": [],
            "configuredProductId": "01t0q000000fuYOAAY",
            "configured": true,
            "configurationData": null,
            "changedByProductActions": false
        }
    ],
    "listPrice": null,
    "isUpgrade": false,
    "isDynamicOption": false,
    "inheritedConfigurationData": null,
    "hiddenOptionIds": null,
    "dynamicOptionKey": null,
    "disabledOptionIds": null,
    "configuredProductId": "01t0q000000ezfNAAQ",
    "configured": false,
    "configurationData": {
        "attributes": {
            "type": "SBQQ__ProductOption__c"
        }
    },
    "changedByProductActions": false
}
```

**Comment by jonathanwiesel** [comment]
@paustint I'm using the `configured = true` on the outermost `configuration` without problem, in fact the error is a good indicative because that means the API is trying to parse the configuration you're passing it but it fails at some point. 

Try replicating as much as you can the attributes that represent a difference, for example:

* `configurationData` create an empty instance of a ProductOption instead of setting it just to `null` (if the API does something like `configurationData.get('X')` it won't fail with a null pointer but simply return `null`).

* The `optionConfigurations` as you said make it an empty array instead of `null` 

* Also, the `configured` attribute on the innermost option your adding manually should be `false` since that option doesn't have more `optionConfigurations` inside itself.

**Comment by paustint** [comment]
@jonathanwiesel - I wanted to close out this thread, as my products FINALLY were added!
I am actually using Apex for this project and then using JS on the side just to more quickly iterate and look at the data.

As far as I can tell (still a bit nebulous) there was a new field added, which was not in the CPQ documentation data model example, `configurationEntered` - I set this to true on all options, including the outermost option and now the products are added to the quote.

Thank you so much for your assistance!
